sgx/common/src/retrieve_key_protocol.c

```diff
@@ -0,0 +1,1164 @@
+}
+
+/*****************************************************************************
+ * parse_key_request_msg()
+ ****************************************************************************/
+int parse_key_request_msg(X509 * client_sign_cert,
+                          ByteBuffer * msg_dec_key,
+                          ECDHMessage * msg_in,
+                          EVP_PKEY * server_eph_pubkey,
+                          ByteBuffer * kmip_request)
+{
+  // decrypt message using input message encryption key
+  ECDHMessage pt_msg = { { 0 }, NULL };
+
+  if (EXIT_SUCCESS != aes_gcm_decrypt(msg_dec_key->buffer,
+                                      msg_dec_key->size,
+                                      msg_in->body,
+                                      msg_in->hdr.msg_size,
+                                      &(pt_msg.body),
+                                      (size_t *) &(pt_msg.hdr.msg_size)))
+  {
+    kmyth_sgx_log(LOG_ERR, "failed to decrypt the 'Key Request' message");
+    if (pt_msg.body != NULL)
+    {
+      free(pt_msg.body);
+    }
+    return EXIT_FAILURE;
+  }
+
+  // parse message body fields into variables
+  int buf_index = 0;
+
+  // get size (in bytes) of KMIP 'get key' request field
+  kmip_request->size = pt_msg.body[buf_index] << 8;
+  kmip_request->size += pt_msg.body[buf_index+1];
+  buf_index += 2;
+
+  // get KMIP 'get key' request bytes
+  kmip_request->buffer = malloc(kmip_request->size);
+  memcpy(kmip_request->buffer, pt_msg.body+buf_index, kmip_request->size);
+  buf_index += kmip_request->size;
+
+  // get size of server-side ephemeral public key field
+  uint16_t server_eph_pub_len = pt_msg.body[buf_index] << 8;
+  server_eph_pub_len += pt_msg.body[buf_index+1];
+  buf_index += 2;
+
+  // get server-side ephemeral public key field bytes
+  unsigned char *server_eph_pub_bytes = malloc(server_eph_pub_len);
+  memcpy(server_eph_pub_bytes, pt_msg.body+buf_index, server_eph_pub_len);
+  buf_index += server_eph_pub_len;
+
+  // buffer index now points just past end of message body
+  // capture this index so we can access message body as part of input buffer
+  size_t msg_body_size = buf_index;
+
+  // get size of message signature
+  uint16_t msg_sig_len = pt_msg.body[buf_index] << 8;
+  msg_sig_len += pt_msg.body[buf_index+1];
+  buf_index += 2;
+
+  // get message signature bytes
+  uint8_t *msg_sig_bytes = malloc(msg_sig_len);
+  memcpy(msg_sig_bytes, pt_msg.body+buf_index, msg_sig_len);
+  buf_index += msg_sig_len;
+
+  // check that number of parsed bytes matches message length input parameter
+  if (buf_index != pt_msg.hdr.msg_size)
+  {
+    kmyth_sgx_log(LOG_ERR, "parsed byte count mismatches input message length");
+    free(server_eph_pub_bytes);
+    free(msg_sig_bytes);
+    return EXIT_FAILURE;
+  }
+
+  // extract server's public signing key (needed to verify signature over
+  // message) from X509 certificate
+  EVP_PKEY *msg_sign_pubkey = X509_get_pubkey(client_sign_cert);
```

Comment:
`msg_sign_pubkey` is never freed.
```
EVP_PKEY_free(msg_sign_pubkey);
```

Reply:
I believe that this issue is addressed by commits [d79c3c2](https://github.com/NationalSecurityAgency/kmyth/pull/168/commits/d79c3c2b85aac1855ade5bd9316f001248452546) and [6571859](https://github.com/NationalSecurityAgency/kmyth/pull/168/commits/6571859f87f01bd6d5c2026783784d65aeae35a3).